src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+  static datalog::ArgumentType GetTypingOrFatal(
+      const TypeEnv& type_env, const std::string& argument_name) {
+    if (type_env.find(argument_name) == type_env.end()) {
+      LOG(FATAL) << "could not find typing for argument" << argument_name;
+    } else {
+      return type_env.find(argument_name)->second;
+    }
+  }
+
+  static datalog::RelationDeclaration GetDeclarationOrFatal(
+      const DeclEnv& decl_env, const std::string& predicate_name) {
+    if (decl_env.find(predicate_name) == decl_env.end()) {
+      LOG(FATAL) << "could not find declaration for relation" << predicate_name;
+    } else {
+      return decl_env.find(predicate_name)->second;
+    }
+  }
+
+  // This generates a set of says assertions that populate all the
+  // universes for all the speakers given the type environment. Essentially,
+  // it loops through the list of constant principals, and then all the
+  // constants of all types, and then generates a SaysAssertion that
+  // adds the constant to the universe for each speaker.
+  static std::vector<SaysAssertion> GetUniverseDefiningFacts(
+      const TypeEnv& type_env) {
+    std::vector<Principal> constant_principals;
+    for (const auto& typing : type_env) {
+      if (typing.second.kind() == datalog::ArgumentType::Kind::kPrincipal) {
+        constant_principals.push_back(Principal(typing.first));
+      }
+    }
+
+    std::vector<SaysAssertion> universe_defining_says_assertions;
+    for (const auto& prin : constant_principals) {
+      for (const auto& typing : type_env) {
+        universe_defining_says_assertions.push_back(SaysAssertion(
+            prin,  // speaker
+            {Assertion(Fact({}, MakeUniverseMembershipFact(typing.first,
+                                                           typing.second)))}));
+      }
+    }
+    return universe_defining_says_assertions;
+  }
+
+  // This visitor populates a mapping (flat_hash_map) of constant (literal)
+  // names to their types by looking at the relations to which the literals
+  // are applied and declarations of those relations (including the types
+  // of the parameters where they are applied). It takes as input
+  // a mapping from relation names to relation declarations.
+  class TypeEnvironmentGenerationVisitor
+      : public AuthLogicAstTraversingVisitor<TypeEnvironmentGenerationVisitor> {
+   public:
+    // A map from the names of principal / predicate argument literals
+    // to typings.
+    using TypeEnv = absl::flat_hash_map<std::string, datalog::ArgumentType>;
+    // A map from predicate names to predicate declarations.
+    using DeclEnv =
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration>;
+
+    TypeEnvironmentGenerationVisitor(
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration> decl_env)
+        : decl_env_(decl_env) {}
+
+    const TypeEnv& type_env() const { return type_env_; }
+
+   private:
+    void AddTyping(std::string arg_name, datalog::ArgumentType arg_type) {
```

Comment:
If you can't, or don't want to, use std::move to ensure efficient copy elision or moving of `arg_name`, use an `absl::string_view` to take the parameter in instead.

Reply:
fixed